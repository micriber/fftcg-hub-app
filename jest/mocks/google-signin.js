export const mockUserInfo = {
    idToken: 'mockIdToken',
    accessToken: null,
    accessTokenExpirationDate: null, // DEPRECATED, on iOS it's a time interval since now in seconds, on Android it's always null
    serverAuthCode: 'mockServerAuthCode',
    scopes: [], // on iOS this is empty array if no additional scopes are defined
    user: {
        email: 'mockEmail',
        id: 'mockId',
        givenName: 'mockGivenName',
        familyName: 'mockFamilyName',
        photo: 'mockPhotoUtl',
        name: 'mockFullName',
    },
};
