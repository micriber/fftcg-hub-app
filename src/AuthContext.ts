import React from 'react';
import {UserInfo} from './services/api/user';

export const AuthContext = React.createContext({
  getCurrentUser: (): UserInfo | null => null,
  getIdToken: (): string | null => null,
  signIn: () => Promise.resolve(),
  signOut: () => Promise.resolve(),
});
