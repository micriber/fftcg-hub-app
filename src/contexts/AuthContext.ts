import React from 'react';
import {UserInfo} from '../services/api/user';

type UserState = {
  isSignedIn: boolean;
  info?: UserInfo;
  idToken?: string;
};

const user: UserState = {isSignedIn: false};
export const AuthContext = React.createContext({
  user,
  isLoading: false,
  getCurrentUser: (): UserInfo | null => null,
  getIdToken: (): string | null => null,
  signIn: () => Promise.resolve(),
  signOut: () => Promise.resolve(),
});
