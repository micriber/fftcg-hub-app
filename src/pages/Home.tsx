import React from 'react';
import {StackNavigationProp} from '@react-navigation/stack';
import {Button, SafeAreaView, ActivityIndicator, Text} from 'react-native';
import {useAsync} from 'react-async';
import {GoogleSignin} from '@react-native-community/google-signin';
import {signOut} from '../services/google';

export type RootStackParamList = {
  Home: undefined;
  About: undefined;
  Login: undefined;
};

type ProfileScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Home'
>;

type Props = {
  navigation: ProfileScreenNavigationProp;
};

const Home = ({navigation}: Props) => {
  const logoutAndRedirect = async () => {
    await signOut();
    navigation.navigate('Login');
  };
  const {data, error, isLoading} = useAsync({
    promiseFn: GoogleSignin.getCurrentUser,
  });
  if (isLoading) {
    return <ActivityIndicator size="small" color="#0000ff" />;
  }
  if (error) {
    return `Something went wrong: ${error.message}`;
  }

  return (
    <SafeAreaView>
      <Text>Welcome {data?.user.name}</Text>
      <Button
        title="Go to about page"
        onPress={() => navigation.navigate('About')}
      />
      <Button title="Logout" onPress={logoutAndRedirect} />
    </SafeAreaView>
  );
};

export default Home;
