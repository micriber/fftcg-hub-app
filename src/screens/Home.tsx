import React from 'react';
import {StackNavigationProp} from '@react-navigation/stack';
import {
  View,
  Image,
  StyleSheet,
  FlatList,
  Dimensions,
  Text,
  TextInput,
} from 'react-native';
import {useAsync} from 'react-async';
import {AuthContext} from '../AuthContext';
import Loading from './Loading';
import cards from '../../jest/mocks/fftcg';
import FastImage from 'react-native-fast-image';

export type RootStackParamList = {
  Home: undefined;
  About: undefined;
  Login: undefined;
};

type ProfileScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Home'
>;

type Props = {
  navigation: ProfileScreenNavigationProp;
};

const w = Dimensions.get('window');

const Home = () => {
  const {signOut, getCurrentUser} = React.useContext(AuthContext);

  const {data, isLoading} = useAsync({
    promiseFn: getCurrentUser,
  });

  if (isLoading) {
    return <Loading />;
  }

  if (!data) {
    signOut();
    return null;
  }

  return (
    <View style={[styles.container]}>
      <TextInput
        blurOnSubmit
        autoCapitalize="none"
        placeholder={'Rechercher "1-001R"'}
        autoCorrect={false}
        style={{
          borderRadius: 25,
          borderColor: '#333',
          backgroundColor: '#fff',
          paddingLeft: 50,
        }}
        // textStyle={{color: '#000'}}
      />
      <View style={[styles.cardContainer]}>
        <FlatList
          initialNumToRender={2}
          numColumns={3}
          data={cards}
          keyExtractor={(item) => item.Code}
          renderItem={({item}) => (
            <View style={{flex: 1, flexDirection: 'column', margin: 1}}>
              <FastImage
                // style={styles.image}
                style={{width: w.width / 4, height: w.height / 4}}
                source={{
                  uri: `https://fftcg.cdn.sewest.net/images/cards/thumbs/${item.Code}_fr.jpg`,
                }}
                resizeMode="contain"
              />
            </View>
          )}
        />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
    marginLeft: 20,
    marginRight: 20,
  },
  cardContainer: {
    // flex: 1,
    marginTop: 3,
    marginLeft: 15,
    // flexDirection: 'row',
    // alignItems: 'flex-start',
    // justifyContent: 'space-between',
  },
  // image: {
  //   height: '25%',
  //   width: '25%',
  // },
});

export default Home;
