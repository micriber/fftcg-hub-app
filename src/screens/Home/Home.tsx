import React from 'react';
import {Alert, Text} from 'react-native';
import useAsync from '../../utils/hooks/useAsync';
import {getCards} from '../../services/api/card';
import {AuthContext} from '../../AuthContext';
import Loading from '../Loading';
import FFCardsList from '../../components/FFCardsList';
import HeaderSwitch from '../../components/common/HeaderSwitch';
import BottomRightButton from '../../components/common/BottomRightButton';

const Home = () => {
  const [isListView, setIsListView] = React.useState(false);
  const {getIdToken} = React.useContext(AuthContext);
  const token = getIdToken();
  const state = useAsync(async () => {
    return await getCards({token: token!, params: {}});
  }, []);

  if (state.loading) {
    return <Loading />;
  }
  if (state.error) {
    return <Text>{JSON.stringify(state.error)}</Text>;
  }
  if (!state.value || (state.value && 'message' in state.value)) {
    return <Text>{JSON.stringify(state.value?.message)}</Text>;
  }

  const cards = state.value.data;

  return (
    <>
      <HeaderSwitch
        leftIconName="appstore-o"
        rightIconName="bars"
        value={isListView}
        onValueChange={setIsListView}
      />
      <FFCardsList isListView={isListView} cards={cards} />
      <BottomRightButton
        iconName="search1"
        onPress={() => Alert.alert('Search button pressed')}
      />
    </>
  );
};

export default Home;
