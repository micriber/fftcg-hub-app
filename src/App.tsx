import * as React from 'react';
import {
  DarkTheme as RNDarkTheme,
  DefaultTheme,
  NavigationContainer,
} from '@react-navigation/native';
import {AppearanceProvider, useColorScheme} from 'react-native-appearance';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import {createStackNavigator} from '@react-navigation/stack';
import {createBottomTabNavigator} from '@react-navigation/bottom-tabs';
import AntIcon from 'react-native-vector-icons/AntDesign';
import {GoogleSignin, statusCodes} from '@react-native-community/google-signin';
import {AuthContext} from './AuthContext';
import About from './screens/About';
import Home from './screens/Home';
import Login from './screens/Login';
import Loading from './screens/Loading';
import Settings from './screens/Settings';
import {signOut} from './services/google';
import {googleLogin, UserInfo} from './services/api/user';
import {Alert} from 'react-native';

interface ITabBarIcon {
  color: string;
  size: number;
}

const AuthStack = createStackNavigator();
const Tabs = createBottomTabNavigator();

const HomeStack = createStackNavigator();
const AboutStack = createStackNavigator();
const SettingsStack = createStackNavigator();

const HomeStackScreen = () => (
  <HomeStack.Navigator>
    <HomeStack.Screen
      options={{headerShown: false}}
      name="Home"
      component={Home}
    />
  </HomeStack.Navigator>
);

const AboutStackScreen = () => (
  <AboutStack.Navigator>
    <AboutStack.Screen
      options={{headerShown: false}}
      name="About"
      component={About}
    />
  </AboutStack.Navigator>
);

const SettingsStackScreen = () => (
  <SettingsStack.Navigator>
    <SettingsStack.Screen
      options={{headerShown: false}}
      name="Settings"
      component={Settings}
    />
  </SettingsStack.Navigator>
);

type UserState = {
  isSignedIn: boolean;
  info?: UserInfo;
  idToken?: string;
};

const App = () => {
  const scheme = useColorScheme();
  const theme = scheme === 'dark' ? RNDarkTheme : DefaultTheme;
  const [isLoading, setIsLoading] = React.useState(true);
  const [user, setUser] = React.useState<UserState>({isSignedIn: false});

  const authContext = React.useMemo(() => {
    return {
      getCurrentUser: () => {
        if (!user.isSignedIn || !user.info) {
          return null;
        }

        return user.info;
      },
      getTokens: async () => {
        return await GoogleSignin.getTokens();
      },
      signIn: async (silently = false) => {
        setIsLoading(true);
        await GoogleSignin.hasPlayServices();
        let userInfo;
        try {
          userInfo = silently
            ? await GoogleSignin.signInSilently()
            : await GoogleSignin.signIn();
        } catch (e) {
          if (e.code === statusCodes.SIGN_IN_REQUIRED) {
            setIsLoading(false);
            return;
          }

          console.log(e);
        }

        const idToken = userInfo?.idToken;
        console.log({userInfo});
        if (!idToken) {
          Alert.alert(
            'Erreur',
            'Impossible de récupérer les informations requises depuis Google.',
          );
          setIsLoading(false);
          return;
        }
        try {
          const signedInUser = await googleLogin(idToken);
          console.log(JSON.stringify(signedInUser, null, 2));
          if ((signedInUser as UserInfo).id) {
            setUser({isSignedIn: true, info: signedInUser as UserInfo});
          }
        } catch (e) {
          Alert.alert(
            'Erreur',
            'Un problème avec votre connexion est survenue. Merci de réessayer ultérieurement.',
          );
        } finally {
          setIsLoading(false);
        }
      },
      signOut: async () => {
        setIsLoading(false);
        setUser({isSignedIn: false});
        await signOut();
      },
    };
  }, [user.info, user.isSignedIn]);

  React.useEffect(() => {
    authContext.signIn(true);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (isLoading) {
    return <Loading />;
  }

  return (
    <AppearanceProvider>
      <SafeAreaProvider>
        <AuthContext.Provider value={authContext}>
          <NavigationContainer theme={theme}>
            {user.isSignedIn ? (
              <Tabs.Navigator
                initialRouteName="Home"
                tabBarOptions={{
                  activeTintColor: theme.colors.primary,
                }}>
                <Tabs.Screen
                  name="About"
                  component={AboutStackScreen}
                  options={{
                    tabBarLabel: 'About',
                    tabBarIcon: ({color, size}: ITabBarIcon) => (
                      <AntIcon name="camera" color={color} size={size} />
                    ),
                  }}
                />
                <Tabs.Screen
                  name="Home"
                  component={HomeStackScreen}
                  options={{
                    tabBarLabel: 'Home',
                    tabBarIcon: ({color, size}: ITabBarIcon) => (
                      <AntIcon name="home" color={color} size={size} />
                    ),
                  }}
                />
                <Tabs.Screen
                  name="Settings"
                  component={SettingsStackScreen}
                  options={{
                    tabBarLabel: 'Settings',
                    tabBarIcon: ({color, size}: ITabBarIcon) => (
                      <AntIcon name="setting" color={color} size={size} />
                    ),
                  }}
                />
              </Tabs.Navigator>
            ) : (
              <AuthStack.Navigator
                screenOptions={{
                  headerShown: false,
                }}>
                <AuthStack.Screen name="Login" component={Login} />
              </AuthStack.Navigator>
            )}
          </NavigationContainer>
        </AuthContext.Provider>
      </SafeAreaProvider>
    </AppearanceProvider>
  );
};

export default App;
