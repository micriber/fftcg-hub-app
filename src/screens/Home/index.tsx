import React from 'react';
import {createStackNavigator, StackHeaderProps} from '@react-navigation/stack';
import Home from './Home';
import Search from './Search';
import CardDetails from '../Cards/CardDetail';
import Header from '../../components/navigation/header';

const HomeStack = createStackNavigator();

const HomeStackScreen = () => (
  <HomeStack.Navigator
    screenOptions={{
      header: (headerProps: StackHeaderProps) => <Header {...headerProps} />,
    }}>
    <HomeStack.Screen
      options={{
        headerTitle: 'Home',
        headerShown: true,
      }}
      name="Home"
      component={Home}
    />
    <HomeStack.Screen name="CollectionSearch" component={Search} />
    <HomeStack.Screen
      name="CardDetails"
      component={CardDetails}
      // TODO: Remove this any
      options={(props) => {
        const {route} = props;
        const pageTitle = route.params && (route.params as any).pageTitle;
        return {
          ...(pageTitle ? {title: pageTitle} : {}),
          headerShown: true,
        };
      }}
    />
  </HomeStack.Navigator>
);

export {HomeStack, HomeStackScreen};
