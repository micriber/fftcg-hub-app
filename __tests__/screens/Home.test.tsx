/**
 * @format
 */

import 'react-native';
import React from 'react';
import Home from '../../src/screens/Home';
// import renderer from 'react-test-renderer';
import {render, waitFor} from '@testing-library/react-native';
import {AuthContext} from '../../src/AuthContext';
import {User} from '@react-native-community/google-signin';

// More about testing in this repository:
// https://github.com/bamlab/react-native-testing/blob/master/src/pages/Home/__tests__/Home.test.tsx

it('renders correctly', async () => {
  const userInfo = {
    user: {
      id: '',
      name: 'Bobby',
      email: 'bobby.billy@gmail.com',
      photo: '',
      familyName: 'Billy',
      givenName: null,
    },
    idToken: null,
    serverAuthCode: null,
  };
  const root = render(
    <AuthContext.Provider
      value={{
        getCurrentUser: (): Promise<User | null> => Promise.resolve(userInfo),
        signIn: () => Promise.resolve(),
        signOut: () => Promise.resolve(),
      }}>
      <Home />
    </AuthContext.Provider>,
  );

  expect(root.toJSON()).toMatchSnapshot(); // Loading screen
  await waitFor(() => root.getByText('Welcome Bobby'));
  expect(root.toJSON()).toMatchSnapshot(); // Home page loaded with user
});
