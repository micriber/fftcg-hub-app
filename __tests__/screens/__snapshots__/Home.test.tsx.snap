// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#0000ff"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`renders correctly 2`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text>
    Welcome 
    Bobby
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
